File: core/src/commonTest/kotlin/sudoku/core/HintTest.kt
package sudoku.core

import sudoku.core.generator.buildHighlights
import sudoku.core.model.*
import sudoku.core.solver.StepFinder
import kotlin.test.*

class HintTest {
    // Easy puzzle solvable with singles
    private val easyPuzzle = "530070000600195000098000060800060003400803001700020006060000280000419005000080079"

    // Harder puzzle requiring advanced techniques
    private val hardPuzzle = "000801000000000043500000000000070800000000100020030000600000075003400000000200600"

    @Test
    fun testStepFinderReturnsStepForValidBoard() {
        val board = Board()
        board.loadFromString(easyPuzzle)
        val finder = StepFinder()
        val step = finder.findNextStep(board)
        assertNotNull(step, "StepFinder should find a step for a valid puzzle")
        assertTrue(step.type.isSingle, "Easy puzzle first step should be a single")
    }

    @Test
    fun testStepFinderReturnsNullForSolvedBoard() {
        // Load a fully completed grid — no empty cells
        val solved = "534678912672195348198342567859761423426853791713924856961537284287419635345286179"
        val board = Board()
        board.loadFromString(solved)
        val finder = StepFinder()
        val step = finder.findNextStep(board)
        assertNull(step, "StepFinder should return null for a solved board")
    }

    @Test
    fun testDescribeVagueReturnsTechniqueName() {
        val step =
            SolutionStep(
                type = SolutionType.NAKED_SINGLE,
                cellIndex = 10,
                value = 5,
            )
        assertEquals("Naked Single", step.describeVague())
    }

    @Test
    fun testDescribeConcreteSingleIncludesCellAndDigit() {
        val step =
            SolutionStep(
                type = SolutionType.HIDDEN_SINGLE,
                cellIndex = 20, // row 3, col 3 (1-indexed: r3c3)
                value = 7,
            )
        val desc = step.describeConcrete()
        assertTrue(desc.contains("Hidden Single"), "Should include technique name")
        assertTrue(desc.contains("7"), "Should include digit")
        assertTrue(desc.contains("r3c3"), "Should include cell position")
    }

    @Test
    fun testDescribeConcreteEliminationIncludesDigits() {
        val step =
            SolutionStep(
                type = SolutionType.NAKED_PAIR,
                indices = listOf(0, 1),
                candidatesRemoved = listOf(2 to 3, 5 to 3, 2 to 7, 5 to 7),
            )
        val desc = step.describeConcrete()
        assertTrue(desc.contains("Naked Pair"), "Should include technique name")
        assertTrue(desc.contains("3") && desc.contains("7"), "Should include eliminated digits")
    }

    @Test
    fun testDescribeConcreteFishIncludesDigit() {
        val step =
            SolutionStep(
                type = SolutionType.X_WING,
                value = 4,
                indices = listOf(0, 8, 72, 80),
                candidatesRemoved = listOf(3 to 4, 6 to 4),
            )
        val desc = step.describeConcrete()
        assertTrue(desc.contains("X-Wing"), "Should include technique name")
        assertTrue(desc.contains("4"), "Should include digit")
    }

    @Test
    fun testBuildHighlightsForSingle() {
        val board = Board()
        board.loadFromString(easyPuzzle)
        val finder = StepFinder()
        val step = finder.findNextStep(board)
        assertNotNull(step)
        val highlights = buildHighlights(board, step)
        assertTrue(highlights.isNotEmpty(), "Highlights should not be empty")
        assertTrue(
            highlights.any { it.role == HighlightRole.DEFINING },
            "Single should have DEFINING highlights",
        )
    }

    @Test
    fun testBuildHighlightsForElimination() {
        val board = Board()
        board.loadFromString(hardPuzzle)
        val finder = StepFinder()
        // Advance past singles to find an elimination step
        var step: SolutionStep?
        do {
            step = finder.findNextStep(board)
            if (step == null) break
            if (!step.type.isSingle) break
            // Apply the step to advance
            board.setCell(step.cellIndex, step.value)
            board.setAllExposedSingles()
        } while (true)

        if (step != null && !step.type.isSingle) {
            val highlights = buildHighlights(board, step)
            assertTrue(highlights.isNotEmpty(), "Elimination step should produce highlights")
            assertTrue(
                highlights.any { it.role == HighlightRole.ELIMINATION },
                "Should have ELIMINATION highlights for ${step.type}",
            )
        }
    }

    @Test
    fun testBuildSolverBoardFromValues() {
        // Simulate what GameViewModel.buildSolverBoard does
        val values = IntArray(81)
        val puzzle = easyPuzzle
        for (i in puzzle.indices) {
            val ch = puzzle[i]
            if (ch in '1'..'9') values[i] = ch - '0'
        }

        val board = Board()
        for (i in 0 until 81) {
            if (values[i] != 0) board.setCell(i, values[i], isFixed = true)
        }

        // Board should have correct values
        for (i in 0 until 81) {
            assertEquals(values[i], board.values[i], "Cell $i should match")
        }

        // Board should have candidates for empty cells
        for (i in 0 until 81) {
            if (values[i] == 0) {
                assertTrue(board.cells[i] != 0, "Empty cell $i should have candidates")
            }
        }

        // StepFinder should find steps on this board
        val finder = StepFinder()
        val step = finder.findNextStep(board)
        assertNotNull(step, "Should find a step on reconstructed board")
    }
}

File: rust/sudoku-core/tests/hint_test.rs
use sudoku_core::board::Board;
use sudoku_core::generator::highlighter::build_highlights;
use sudoku_core::solver::StepFinder;
use sudoku_core::step::SolutionStep;
use sudoku_core::types::{HighlightRole, SolutionType};

const EASY_PUZZLE: &str = "530070000600195000098000060800060003400803001700020006060000280000419005000080079";
const HARD_PUZZLE: &str = "000801000000000043500000000000070800000000100020030000600000075003400000000200600";

#[test]
fn test_step_finder_returns_step_for_valid_board() {
    let mut board = Board::new();
    board.load_from_string(EASY_PUZZLE);
    let finder = StepFinder::new();
    let step = finder.find_next_step(&board, None);
    assert!(step.is_some(), "StepFinder should find a step for a valid puzzle");
    assert!(step.unwrap().step_type.is_single(), "Easy puzzle first step should be a single");
}

#[test]
fn test_step_finder_returns_none_for_solved_board() {
    let solved = "534678912672195348198342567859761423426853791713924856961537284287419635345286179";
    let mut board = Board::new();
    board.load_from_string(solved);
    let finder = StepFinder::new();
    let step = finder.find_next_step(&board, None);
    assert!(step.is_none(), "StepFinder should return None for a solved board");
}

#[test]
fn test_describe_vague() {
    let step = SolutionStep::single(SolutionType::NakedSingle, 10, 5);
    assert_eq!(step.describe_vague(), "Naked Single");
}

#[test]
fn test_describe_concrete_single() {
    let step = SolutionStep::single(SolutionType::HiddenSingle, 20, 7);
    let desc = step.describe_concrete();
    assert!(desc.contains("Hidden Single"));
    assert!(desc.contains("7"));
    assert!(desc.contains("r3c3"));
}

#[test]
fn test_describe_concrete_elimination() {
    let step = SolutionStep::elimination(
        SolutionType::NakedPair,
        vec![0, 1],
        0,
        vec![(2, 3), (5, 3), (2, 7), (5, 7)],
    );
    let desc = step.describe_concrete();
    assert!(desc.contains("Naked Pair"));
    assert!(desc.contains("3") && desc.contains("7"));
}

#[test]
fn test_describe_concrete_fish() {
    let step = SolutionStep::elimination(
        SolutionType::XWing,
        vec![0, 8, 72, 80],
        4,
        vec![(3, 4), (6, 4)],
    );
    let desc = step.describe_concrete();
    assert!(desc.contains("X-Wing"));
    assert!(desc.contains("4"));
}

#[test]
fn test_build_highlights_for_single() {
    let mut board = Board::new();
    board.load_from_string(EASY_PUZZLE);
    let finder = StepFinder::new();
    let step = finder.find_next_step(&board, None).unwrap();
    let highlights = build_highlights(&board, &step);
    assert!(!highlights.is_empty());
    assert!(highlights.iter().any(|h| h.role == HighlightRole::Defining));
}

#[test]
fn test_build_highlights_for_elimination() {
    let mut board = Board::new();
    board.load_from_string(HARD_PUZZLE);
    let finder = StepFinder::new();

    // Advance past singles
    let mut step_opt;
    loop {
        step_opt = finder.find_next_step(&board, None);
        match &step_opt {
            None => break,
            Some(step) if !step.step_type.is_single() => break,
            Some(step) => {
                board.set_cell(step.cell_index as usize, step.value, false);
                board.set_all_exposed_singles();
            }
        }
    }

    if let Some(step) = step_opt {
        if !step.step_type.is_single() {
            let highlights = build_highlights(&board, &step);
            assert!(!highlights.is_empty(), "Elimination step should produce highlights");
            assert!(
                highlights.iter().any(|h| h.role == HighlightRole::Elimination),
                "Should have ELIMINATION highlights for {:?}",
                step.step_type
            );
        }
    }
}

#[test]
fn test_build_solver_board_from_values() {
    let puzzle = EASY_PUZZLE;
    let mut values = [0u8; 81];
    for (i, ch) in puzzle.chars().enumerate() {
        if ch >= '1' && ch <= '9' {
            values[i] = ch as u8 - b'0';
        }
    }

    let mut board = Board::new();
    for i in 0..81 {
        if values[i] != 0 {
            board.set_cell(i, values[i], true);
        }
    }

    for i in 0..81 {
        assert_eq!(board.values[i], values[i], "Cell {} should match", i);
    }
    for i in 0..81 {
        if values[i] == 0 {
            assert_ne!(board.candidates[i], 0, "Empty cell {} should have candidates", i);
        }
    }

    let finder = StepFinder::new();
    assert!(finder.find_next_step(&board, None).is_some());
}
